### PR TITLE
Use `sf::Vector2<T>` for numeric parameter pairs

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

This PR further expands on the work from #1902, #1942, and #1948. It originated from a conversation in #2027 which found more instances of pairs of numeric parameters. I thought it'd be good to go ahead and fix this but quickly found just how many more numeric parameter pairs we had left in SFML. I tried to find as many instances of this pattern as possible but if any still remain, I'd love to fix them I this PR.

I know this PR is massive. While finding and changing all these interfaces, I didn't try to break this apart into multiple commits although it's not unreasonable to do so. If this is simply an unmanageably large PR to review then let me know and I'll try to break it apart.

## Tasks

* [x] Tested on Linux
* [x] Tested on Windows
* [x] Tested on macOS
* [x] Tested on iOS
* [x] Tested on Android

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: include/SFML/Graphics/RenderTexture.hpp [hunk 1]
```diff
@@ -78,14 +78,13 @@ class SFML_GRAPHICS_API RenderTexture : public RenderTarget
...
-    /// \param width    Width of the render-texture
-    /// \param height   Height of the render-texture
+    /// \param size     Width and height of the render-texture
     /// \param settings Additional settings for the underlying OpenGL texture and context
     ///
     /// \return True if creation has been successful
     ///
     ////////////////////////////////////////////////////////////
-    [[nodiscard]] bool create(unsigned int width, unsigned int height, const ContextSettings& settings = ContextSettings());
+    [[nodiscard]] bool create(const sf::Vector2u& size, const ContextSettings& settings = ContextSettings());
```
Comment (JonnyPtn) [inline include/SFML/Graphics/RenderTexture.hpp:87]: I think sf namespace is unnecessary here?

File: include/SFML/Window/Context.hpp [hunk 2]
```diff
@@ -30,6 +30,7 @@
 ////////////////////////////////////////////////////////////
 #include <SFML/Window/Export.hpp>
 #include <SFML/Window/GlResource.hpp>
+#include <SFML/System/Vector2.hpp>
```
Comment (JonnyPtn) [inline include/SFML/Window/Context.hpp:33]: Can we forward declare instead?

**Review by JonnyPtn — CHANGES_REQUESTED** [review]

File: include/SFML/Graphics/Texture.hpp [hunk 3]
```diff
@@ -272,13 +271,11 @@ class SFML_GRAPHICS_API Texture : GlResource
     /// texture was not previously created.
     ///
     /// \param pixels Array of pixels to copy to the texture
-    /// \param width  Width of the pixel region contained in \a pixels
-    /// \param height Height of the pixel region contained in \a pixels
-    /// \param x      X offset in the texture where to copy the source pixels
-    /// \param y      Y offset in the texture where to copy the source pixels
+    /// \param size   Width and height of the pixel region contained in \a pixels
+    /// \param offset Offset in the texture where to copy the source pixels
```
Comment (Bromeon) [inline include/SFML/Graphics/Texture.hpp:275]: > Offset in the texture where to copy the source pixels

This sounds confusing (I know it was before). Is this source or destination?
Maybe we should also rename parameters for clarity.

File: src/SFML/Window/Win32/WindowImplWin32.cpp [hunk 4]
```diff
@@ -226,7 +226,7 @@ m_cursorGrabbed   (m_fullscreen)
 
     // By default, the OS limits the size of the window the the desktop size,
     // we have to resize it after creation to apply the real size
-    setSize(Vector2u(mode.width, mode.height));
+    setSize(Vector2u(mode.size));
```
Comment (kimci86) [inline src/SFML/Window/Win32/WindowImplWin32.cpp:229]:
```suggestion
    setSize(mode.size);
```

File: include/SFML/Window/VideoMode.hpp [hunk 5]
```diff
@@ -53,12 +54,11 @@ class SFML_WINDOW_API VideoMode
...
     /// \brief Construct the video mode with its attributes
     ///
-    /// \param modeWidth        Width in pixels
-    /// \param modeHeight       Height in pixels
+    /// \param modeSize         Width and height in pixels
     /// \param modeBitsPerPixel Pixel depths in bits per pixel
     ///
     ////////////////////////////////////////////////////////////
-    VideoMode(unsigned int modeWidth, unsigned int modeHeight, unsigned int modeBitsPerPixel = 32);
+    VideoMode(const Vector2u& modeSize, unsigned int modeBitsPerPixel = 32);
```
Comment (kimci86) [inline include/SFML/Window/VideoMode.hpp:61]: Note that this is now allowed:
```c++
sf::VideoMode mode = sf::Vector2u(800, 600);
```
Should this constructor be `explicit`?

**Review by kimci86 — COMMENTED** [review]
I have some comments and suggestions below.
In addition, `View::move(float, float)` method could be removed.

File: include/SFML/Window/VideoMode.hpp (hunk 5 above)
Comment (Bromeon) [reply include/SFML/Window/VideoMode.hpp:61]: Good catch -- this also means that we will have silent code breaking from SFML 2 -> 3 (code still compiles, but causes nasty bugs at runtime).

File: include/SFML/Window/Context.hpp (hunk 2 above)
Comment (ChrisThrasher) [reply include/SFML/Window/Context.hpp:33]: Is that possible given how `Vector2` is a class template and `Vector2u` a type alias? I just tried this locally and couldn't get it to work.

File: src/SFML/Window/Win32/WindowImplWin32.cpp (hunk 4 above)
Comment (ChrisThrasher) [reply src/SFML/Window/Win32/WindowImplWin32.cpp]: Thanks!

**Comment by ChrisThrasher** [comment]
> In addition, `View::move(float, float)` method could be removed.

@kimci86 Good catch. The definition for that was already deleted but the declaration stuck around for some reason.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: include/SFML/Window/VideoMode.hpp (hunk 5 above)
Comment (ChrisThrasher) [reply include/SFML/Window/VideoMode.hpp]: Okay! I just fixed it.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
```cpp
    struct SizeEvent
    {
        unsigned int width;  //!< New width, in pixels
        unsigned int height; //!< New height, in pixels
    };
```
Should `sf::Event::SizeEvent` contain a `Vector2u` as well? I was unsure if this was a good idea so I didn't change it.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by Bromeon** [comment]
> Should `sf::Event::SizeEvent` contain a `Vector2u` as well? I was unsure if this was a good idea so I didn't change it.

I remember this was originally avoided because in C++98, unions were not allowed to store types with constructors (see e.g. [here](https://en.sfml-dev.org/forums/index.php?topic=18399). I'm not fully up-to-date; this might have been relaxed in recent standards (as indicated [here](https://stackoverflow.com/q/10693913), for example). Someone would need to confirm this.

If yes, this change can also be applied to the other events which hold a position (mouse, touch).

**Comment by ChrisThrasher** [comment]
> I remember this was originally avoided because in C++98, unions were not allowed to store types with constructors (see e.g. [here](https://en.sfml-dev.org/forums/index.php?topic=18399)). I'm not fully up-to-date; this might have been relaxed in recent standards (as indicated [here](https://stackoverflow.com/q/10693913), for example). Someone would need to confirm this.

AppleClang didn't like it when I tried to do this. Maybe I'm doing something wrong or maybe it really isn't possible, sadly.

```
/Users/thrasher/Projects/sfml/src/SFML/Window/WindowImpl.cpp:196:19: error: call to implicitly-deleted default constructor of 'sf::Event'
            Event event;
```

**Comment by Bromeon** [comment]
> One downside is the the API starts to look weird when using these events. `event.size.width` becomes `event.size.size.x` which frankly looks bad. I'm on the fence about whether this is a worthwhile change to make.

I think the real problem is the naming of `SizeEvent` and `size` (the event). Should probably be `Resize`; then `event.resize.size.x` or `event.resized.size.x` would also look less problematic.

> It can certainly wait for a later PR since it's a bit outside the scope of this PR which aims to only change APIs using x/y or width/height pairs.

Yes, I'd say let's discuss the `sf::Event` API separately.

File: include/SFML/Graphics/Texture.hpp (hunk 3 above)
Comment (ChrisThrasher) [reply include/SFML/Graphics/Texture.hpp:275]: Reading through the source, `offset` appears to representation a destination, not a source.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
> One possibility would also be to look into std::variant -- but the visiting API is quite unfortunate in C++, even [the tricks](https://stackoverflow.com/a/62344801) attempting to make it more ergonomic aren't ideal.

Just as a side-note - and I agree with the separate discussion for the Event API - @lethal-guitar gave a talk on using `std::variant` with SFML's event system at Meeting C++ 2018: https://www.youtube.com/watch?v=CELWr9roNno

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: include/SFML/Window/Context.hpp (hunk 2 above)
Comment (eXpl0it3r) [reply include/SFML/Window/Context.hpp:33]: Yeah, you can't since it's a template

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by Bromeon** [comment]
It's very hard to keep track of PRs which last for a month or more and change 60+ files, especially when small changes are continuously added by force-pushing/overwriting 1 commit. As a reviewer, it would require me to repeatedly re-review everything and memorize the changes from a week ago.

I would suggest we keep PRs smaller and merge them quicker, aka "don't let perfect be the enemy of good" -- we can iteratively improve the codebase without getting everything done at once. In addition to easier reviewing, this would be more aligned with the continous integration philosophy and would also avoid merge conflicts as other changes get merged in the meantime.

Keeping separate commits (until final squash) is an option too, but it only addresses half of the issues.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: include/SFML/Graphics/Texture.hpp (hunk 3 above)
Comment (eXpl0it3r) [reply include/SFML/Graphics/Texture.hpp:275]: Should it be renamed to destination everywhere?

Comment (ChrisThrasher) [reply include/SFML/Graphics/Texture.hpp:275]: I like that idea

**Review by eXpl0it3r — APPROVED** [review]
Looks good to me!

Thanks for the massive effort 🙂 

Before we merge this, we should however make sure, that it really is an intentional sematic change from (width, height) to (size.x, size.y).

I've been using the vector class like this for a long time and thus am totally fine with the change. What do others think about it?

**Comment by Bonkt** [comment]
I was recently very confused as i rebuilt my SFML and had used pairs instead of sf::VectorT for some of my initialization, and couldn't find the overloaded methods in SFML GH repo, nor anything in the change log.
What made it worse was that the documentation (2.5.1) is still stating support for numeric pairs. 

Although i will admit i might have misunderstood what version the "master" Branch here on GH actually is. Is it 2.5.1 or some newer? Maybe i should have cloned an older  version? Sorry - i'm new to open source GH.

Also - sorry if this is an inopportune comment in this forum.

**Comment by Bromeon** [comment]
> Although i will admit i might have misunderstood what version the "master" Branch here on GH actually is. Is it 2.5.1 or some newer? Maybe i should have cloned an older version? Sorry - i'm new to open source GH.

A bit off-topic, but we actively maintain 2 branches on GitHub:
* `master` -- follows SFML 3 development
* `2.6.x` -- continuation of SFML 2 for some time

So yes, `master` branch will not be in line with the documentation you find on the website. You'll need to generate the docs for SFML 3 yourself with Doxygen, at the moment.

**Comment by ChrisThrasher** [comment]
Rebasing on master revealed some compilation errors with DRM code, so I made the same `width, height` -> `size` changes and everything appears to be working now.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Window/DRM/DRMContext.cpp [hunk 6]
```diff
@@ -204,8 +204,7 @@ m_config     (NULL),
 m_currentBO  (NULL),
 m_nextBO     (NULL),
 m_gbmSurface (NULL),
-m_width      (0),
-m_height     (0),
+m_size       (0, 0),
```
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:207]: Do we care about explicitly initializing vectors to zero when the default constructor already does this?

**Comment by ChrisThrasher** [comment]
I improved some parameter names and rebased on master since there were minor conflicts to fix.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]

**Comment by eXpl0it3r** [comment]
Thank you for this massive PR! ❤️🙂 

Lots of breaking changes to document 😅

**Pull request merged**